### PR TITLE
expression: Support more types when getting default flen and decimal

### PR DESCRIPTION
**Pull request by breeswish**
Fix https://github.com/pingcap/tidb/issues/4233

This PR needs at least 3 LGTM to merge.

- Added more default flen / decimal.
- Fixed incorrect REVERSE() return type.
- Fixed incorrect ABS() return type for FLOAT / DOUBLE parameter.
- Fixed incorrect IF / IFNULL type inference: when the decimal of one of the parameter is unspecified, the return decimal should be unspecified.
- Fixed incorrect type inference for some functions: when one of the parameter is a kind of string, its decimal should be treated as unspecified.
- Fixed incorrect decimal / flen in CREATE TABLE.
- Unified default decimal / flen in CAST.
- Fixed `NULLIF(param1, param2)` inconsistent behavior with `IF(param1 = param2, NULL, param1)`.
- Fixed incorrect field name printing:
   - For int types like INT / TINY / BIT and char types like CHAR / VARCHAR, the flen is always printed.
   - For FLOAT / DOUBLE, the flen and decimal is printed only when decimal is specified.
   - For TEXT / BLOB / ..., the flen and decimal is NEVER printed.
- Fixed a bunch of incorrect flen / decimal in typeinfer_test.

~~TODO: We have a bunch of tests to fix.~~ Fixed

~~TODO: might need to fix mysql-test.~~ Fixed

---

After merging this PR, there are still two primary issues causing our FLEN / DECIMAL inference being incorrect:

1. We do not differentiate "significant digit" and "display length" previously. For example, when creating table using `DECIMAL(10, 3)`, the sigificant digit is 10 and the display length is 12 (10 sigificant digit + 1 sign + 1 dot). They are totally mixed together and both of them are stored in `Flen`. Even worse, some of the code treat them as "significant digit" and other as "display length". Other types like `DATETIME`, `FLOAT`, `DOUBLE` have this issue as well.

2. We do not assign correct Flen / Decimal in `types.NewFieldType` (we set both of them to -1) so that during the calculation Flen / Decimal might be incorrect.

### CODE REVIEW

File: mysql/util.go [hunk 1]
```diff
@@ -13,54 +13,43 @@
...
-	switch tp {
-	case TypeNewDecimal:
-		// See https://dev.mysql.com/doc/refman/5.7/en/fixed-point-types.html
-		return 0
-	case TypeDatetime, TypeDuration:
-		return 0
-	default:
-		//TODO: Add more types.
-		return -1
+func GetDefaultFieldLengthAndDecimal(tp byte) (flen int, decimal int) {
```
Comment (XuHuaiyu) [inline mysql/util.go]: can we move this to another file?
this file seems is only used for this?

File: parser/parser.y [hunk 2]
```diff
@@ -4144,12 +4144,6 @@ CastType:
 		x := types.NewFieldType(mysql.TypeNewDecimal)
 		x.Flen = fopt.Flen
 		x.Decimal = fopt.Decimal
-		if fopt.Flen == types.UnspecifiedLength {
```
Comment (XuHuaiyu) [inline parser/parser.y]: Have you checked other types?
If we set the default length and decimal of them in `parser.y`,
can the code be removed?

Comment (breeswish) [inline parser/parser.y]: Good point. Those code should be removed.

File: ddl/ddl_api.go [hunk 3]
```diff
@@ -202,12 +202,13 @@ func setCharsetCollationFlenDecimal(tp *types.FieldType) error {
 			}
 		}
 	}
-	// If flen is not assigned, assigned it by type.
+	// If flen or decimal is not assigned, assigned it by type.
```
Comment (zz-jason) [inline ddl/ddl_api.go]: maybe this comment is more explicit: "use default value for Flen and Decimal when they are unspecified"

File: infoschema/tables.go [hunk 4]
```diff
@@ -679,12 +679,13 @@ func dataForColumnsInTable(schema *model.DBInfo, tbl *model.TableInfo) [][]types
 	rows := [][]types.Datum{}
 	for i, col := range tbl.Columns {
 		colLen := col.Flen
+		decimal := col.Decimal
```
Comment (zz-jason) [inline infoschema/tables.go]:
```go
colLen, colDecimal := col.Flen, col.Decimal
```

File: mysql/util.go (hunk 1 above)
Comment (zz-jason) [inline mysql/util.go]: how about move it to package "types" ?

File: mysql/util.go [hunk 5]
```diff
@@ -13,54 +13,43 @@
 
 package mysql
 
-// GetDefaultFieldLength is used for Interger Types, Flen is the display length.
+var defaultLengthAndDecimal = map[byte]struct {
```
Comment (zz-jason) [inline mysql/util.go]: how about put this map to file "mysql/const.go" ?

File: parser/parser.y (hunk 2 above)
Comment (zz-jason) [inline parser/parser.y]:
```go
x.Flen, x.Decimal = fopt.Flen, fopt.Decimal
```

**Comment by breeswish** [comment]
@zz-jason @shenli @XuHuaiyu @winkyao @jackysp PTAL

**Comment by shenli** [comment]
Can we remove the WIP tag?

**Comment by breeswish** [comment]
@shenli tidb-test is not fixed yet. After that WIP tag could be removed.

File: expression/builtin_miscellaneous.go [hunk 6]
```diff
@@ -199,12 +199,14 @@ func (c *anyValueFunctionClass) getFunction(args []Expression, ctx context.Conte
...
 		sig = &builtinDurationAnyValueSig{baseDurationBuiltinFunc{bf}}
 	case tpInt:
+		bf.tp.Decimal = 0
 		sig = &builtinIntAnyValueSig{baseIntBuiltinFunc{bf}}
 	case tpJSON:
 		sig = &builtinJSONAnyValueSig{baseJSONBuiltinFunc{bf}}
 	case tpReal:
 		sig = &builtinRealAnyValueSig{baseRealBuiltinFunc{bf}}
 	case tpString:
+		bf.tp.Decimal = types.UnspecifiedLength
```
Comment (zz-jason) [inline expression/builtin_miscellaneous.go]: does all the functions which have `tpString` result type should set there `Decimal`s to `types.UnspecifiedLength` ?

Comment (breeswish) [inline expression/builtin_miscellaneous.go]: @zz-jason We need more investigate about this. Known functions compatible with this rules are LOWER, UPPER, REVERSE, IF, ANYVALUE, ...

I guess that, when tpString is used as a parameter, its decimal is set to -1. I will add more examples about this behavior later.

File: executor/ddl_test.go [hunk 7]
```diff
@@ -206,7 +206,8 @@ func (s *testSuite) TestAlterTableModifyColumn(c *C) {
 	result := tk.MustQuery("show create table mc")
 	createSQL := result.Rows()[0][1]
 	// FIXME: `c2` ought to be text, not text(65535).
-	expected := "CREATE TABLE `mc` (\n  `c1` bigint(21) DEFAULT NULL,\n  `c2` text(65535) DEFAULT NULL\n) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin"
+	// FIXME: `c1` should not have the decimal part.
```
Comment (XuHuaiyu) [inline executor/ddl_test.go]: It seems that `c1` has no decimal part?

File: parser/parser.y [hunk 8]
```diff
@@ -2928,9 +2928,17 @@ FunctionCallKeyword:
 	"CAST" '(' Expression "AS" CastType ')'
 	{
 		/* See https://dev.mysql.com/doc/refman/5.7/en/cast-functions.html#function_cast */
+		tp := $5.(*types.FieldType)
+		defaultFlen, defaultDecimal := mysql.GetDefaultFieldLengthAndDecimalForCast(tp.Tp)
+		if tp.Flen == -1 {
```
Comment (XuHuaiyu) [inline parser/parser.y]: s/ -1 / types.UnspecifiedLength ?
so as the others.

File: plan/planbuilder.go [hunk 9]
```diff
@@ -1051,23 +1051,28 @@ func composeShowSchema(names []string, ftypes []byte) *expression.Schema {
 		col := &expression.Column{
 			ColName: model.NewCIStr(name),
 		}
-		var retType types.FieldType
+		var retTp byte
```
Comment (XuHuaiyu) [inline plan/planbuilder.go]: retTp and retType is confused...
can we only use retType

File: util/types/field_type.go [hunk 10]
```diff
@@ -186,6 +186,21 @@ func (ft *FieldType) Init(tp byte) {
 func (ft *FieldType) CompactStr() string {
 	ts := TypeToStr(ft.Tp, ft.Charset)
 	suffix := ""
+
+	defaultFlen, defaultDecimal := mysql.GetDefaultFieldLengthAndDecimal(ft.Tp)
+	isFlenNotDefault := ft.Flen != defaultFlen && ft.Flen != 0 && ft.Flen != UnspecifiedLength
+	isDecimalNotDefault := ft.Decimal != defaultDecimal && ft.Decimal != 0 && ft.Decimal != UnspecifiedLength
+
+	// displayFlen and displayDecimal are flen and decimal values with `-1` substituted with default value.
+	displayFlen := ft.Flen
```
Comment (XuHuaiyu) [inline util/types/field_type.go]: displayFlen, displayDecimal := ft.Flen, ft.Decimal

File: executor/ddl_test.go (hunk 7 above)
Comment (breeswish) [inline executor/ddl_test.go]: Thanks. The comments should be removed.

File: plan/planbuilder.go (hunk 9 above)
Comment (breeswish) [inline plan/planbuilder.go]: 😂😂 retTp is `FieldType.Tp`, and retType is `FieldType`. retTp and retType itself is not the source of confusing.. We may refine the naming later..

Comment (ngaut) [inline plan/planbuilder.go]: Could you please add a issue to track it ?

Comment (breeswish) [inline plan/planbuilder.go]: @ngaut Done! https://github.com/pingcap/tidb/issues/4310

**Comment by zz-jason** [comment]
@XuHuaiyu PTAL

**Comment by XuHuaiyu** [comment]
PTAL @jackysp

**Comment by XuHuaiyu** [comment]
please resolve the conflicts.

**Comment by breeswish** [comment]
@XuHuaiyu Done

**Pull request merged**